contracts/wallet/WalletFactory.sol

```diff
@@ -34,60 +38,161 @@ contract WalletFactory is Owned, Managed {
         address payable wallet = address(proxy);
-        // check for ENS
-        bytes memory labelBytes = bytes(_label);
-        if (labelBytes.length != 0) {
-            // add the factory to the modules so it can claim the reverse ENS
-            address[] memory extendedModules = new address[](_modules.length + 1);
-            extendedModules[0] = address(this);
-            for(uint i = 0; i < _modules.length; i++) {
-                extendedModules[i + 1] = _modules[i];
-            }
-            // initialise the wallet with the owner and the extended modules
-            BaseWallet(wallet).init(_owner, extendedModules);
-            // register ENS
-            registerWalletENS(wallet, _label);
-            // remove the factory from the authorised modules
-            BaseWallet(wallet).authoriseModule(address(this), false);
-        } else {
-            // initialise the wallet with the owner and the modules
-            BaseWallet(wallet).init(_owner, _modules);
+        configureWallet(BaseWallet(wallet), _owner, _modules, _label, address(0));
+    }
+
+    /**
+     * @dev Lets the manager create a wallet for an owner account.
+     * The wallet is initialised with a list of modules, a first guardian, and an optional ENS..
+     * The wallet is created using the CREATE opcode.
+     * @param _owner The account address.
+     * @param _modules The list of modules.
+     * @param _label (Optional) ENS label of the new wallet, e.g. franck.
+     * @param _guardian The guardian address.
+     */
+    function createWalletWithGuardian(
+        address _owner,
+        address[] calldata _modules,
+        string calldata _label,
+        address _guardian
+    )
+        external
+        onlyManager
+    {
+        Proxy proxy = new Proxy(walletImplementation);
+        address payable wallet = address(proxy);
+        configureWallet(BaseWallet(wallet), _owner, _modules, _label, _guardian);
+    }
+
+    /**
+     * @dev Lets the manager create a wallet for an owner account at a specific address.
+     * The wallet is initialised with a list of modules and an optional ENS.
+     * The wallet is created using the CREATE2 opcode.
+     * @param _owner The account address.
+     * @param _modules The list of modules.
+     * @param _label (Optional) ENS label of the new wallet, e.g. franck.
+     * @param _salt The salt.
+     */
+    function createCounterfactualWallet(
+        address _owner,
+        address[] calldata _modules,
+        string calldata _label,
+        bytes32 _salt
+    )
+        external
+        onlyManager
+    {
+        bytes32 newsalt = keccak256(abi.encodePacked(_salt, _owner, _modules));
+        bytes memory code = abi.encodePacked(type(Proxy).creationCode, uint256(walletImplementation));
+        address payable wallet;
+        // solium-disable-next-line security/no-inline-assembly
+        assembly {
+            wallet := create2(0, add(code, 0x20), mload(code), newsalt)
```

Comment:
Since 0.6.2 there is no need to drop to the assembly level https://solidity.readthedocs.io/en/v0.6.2/control-structures.html#salted-contract-creations-create2 as we can call `new` contract with `CREATE2` in Solidity itself.

Reply:
The current pragma version on our contracts is 0.5.4. so we should not use 0.6.x features right?

Reply:
0.5.4 is a year old version and we should move to 0.6.x at some point. This is a worthwhile feature to push us to upgrade.

Reply:
Agree that we should migrate to 0.6.x at some point but let's do it after the next contract deployment (ENS migration and this PR).